test/test_webrtc_connection_mixin.py

```diff
@@ -0,0 +1,272 @@
+        with pytest.raises(ValueError, match=r"Unsupported track kind .*"):
+            await self.send_offer(pc, test_client, audio=audio, video=True)
+        await self.close_peer_connection(pc)
+
+    @pytest.mark.asyncio
+    @pytest.mark.parametrize("test_client_and_stream", [{"modality": "video"}],  indirect=True)
+    @pytest.mark.parametrize("video", [True, False])
+    async def test_unsuccessful_connection_video(self, test_client_and_stream, video):
+        test_client, stream = test_client_and_stream
+        pc, channel = await self.setup_peer_connection(audio=True, video=video)
+        with pytest.raises(ValueError, match=r"Unsupported track kind .*"):
+            await self.send_offer(pc, test_client, audio=True, video=video)
+        await self.close_peer_connection(pc)
+
+    @pytest.mark.asyncio
+    async def test_unsuccessful_webrtc_offer_no_webrtc_id(self, test_client_and_stream):
+        test_client, stream = test_client_and_stream
+        pc, channel = await self.setup_peer_connection()
+        await self.send_offer(
+            pc,
+            test_client,
+            webrtc_id=None,
+            response_code=422,
+        )
+
+    @pytest.mark.asyncio
+    @pytest.mark.parametrize("test_client_and_stream", [{"modality": "dummy"}],  indirect=True)
+    @pytest.mark.parametrize("audio, video", [
+        (True, False),
+        (False, True),
+        (True, True),
+    ])
+    async def test_incorrect_modality(self, test_client_and_stream, audio, video):
+        test_client, stream = test_client_and_stream
+        pc, channel = await self.setup_peer_connection(audio=audio, video=video)
+        with pytest.raises(ValueError, match=r"Modality must be .*"):
+            await self.send_offer(pc, test_client, audio=audio, video=video)
+        await self.close_peer_connection(pc)
+
+    @pytest.mark.asyncio
+    @pytest.mark.parametrize("test_client_and_stream", [{"concurrency_limit": 1}], indirect=True)
+    async def test_concurrency_limit_reached_two_peers(self, test_client_and_stream):
+        test_client, stream = test_client_and_stream
+        pc1, channel = await self.setup_peer_connection(video=True)
+        pc2, channel = await self.setup_peer_connection(video=True)
+        await self.send_offer(pc1, test_client)
+        status, metadata = await self.send_offer(pc2, test_client, return_status_and_metadata=True)
+
+        assert status == 'failed'
+        assert metadata == {'error': 'concurrency_limit_reached', 'limit': 1}
+
+        await self.close_peer_connection(pc1)
+        await self.close_peer_connection(pc2)
+
+    @pytest.mark.asyncio
+    @pytest.mark.parametrize("test_client_and_stream", [{"concurrency_limit": 2}], indirect=True)
+    @pytest.mark.xfail(reason="It is the same webrtc_id so pc1 and pc2 are seen as ONE connection")
+    async def test_concurrency_limit_reached_three_peers_same_id(self, test_client_and_stream):
```

Comment:
Here, as long as ``concurrency_limit>1`` and you reuse ``webrtc_id`` you can supply as many peer connections you want since ``self.connections`` has length 1.

Should probably be restricted in some way, alternatively have a limit for how many connections one ``webrtc_id`` can have.

Reply:
We should not allow duplicate webrtc_ids. Will push in your branch.